### DISCUSSION TITLE
Maintain export, but ignore missing?

### DISCUSSION THREAD

**Discussion by DanielHenriksen** [RhetTbull/osxphotos#357] [General]
Hi! Is it possible to programmatically have osxphotos maintain an export, but ignore previously exported files (but missing)?

Case: I testing Eagle (https://eagle.cool) and got them to include an auto-import option. It watches a designated folder and then imports new files to Eagle. But it moves the files instead of copying them - making osxphotos export them again.

It makes sense to move the files to avoid duplicates, but can osxphotos still update the export with changed/new - ignoring previously exported files?

Daniel :)

**Comment by RhetTbull** [comment]
Hi @DanielHenriksen.  I understand the use case but currently there's no way to avoid re-exporting the images with osxphotos.  There is a "--update" option which I presume you're using to only export updated or missing files but osxphotos would see the files are missing and re-export.  

It would be possible to implement this behavior but would require rewrite of the --update code to add an option (maybe --ignore-missing) to do this.

One possibility would be to use the "--from-date" option to only export images added to Photos since a certain date.  For example:

`osxphotos export /path/to/export --update --from-date 2021-02-01`

--from-date uses ISO 8601 format and can take any of these forms:

- 2000-01-12T12:00:00 (with time)
- 2001-01-12T12:00:00-07:00 (with time and timezone offset)
- 2000-12-31 (no time, assumes midnight)

**Comment by RhetTbull** [comment]
I've taken a look at the code and I think this would be easier to implement than I originally guessed.  I've added a new issue for this (#358).  Not sure exactly when I'll get to this as I'm currently working on another big feature but I'll eventually get around to it.

**Reply by DanielHenriksen** [reply]
Thank you :)

**Reply by RhetTbull** [reply]
This turned out to be much easier than I first thought so I went ahead and implemented it in [v0.40.5](https://github.com/RhetTbull/osxphotos/releases/tag/v0.40.5) :-)

I've not done extensive testing yet so let me know if you encounter any issues.  

With this release, you can do this:

`osxphotos export /path/to/export --update --only-new` 

and only photos which had previously not been exported will be exported.

**Comment by DanielHenriksen** [comment]
Hello! FYI: I'm getting a lot of these now. Not sure if it's related ...
(Also exiftool-errors, but not osxphotos related. They just weren't there before.)

Export-log:

`Fre 5 Feb 2021 16:00:00 UTC
Exporting 2268 photos to /Users/dsh/Pictures/Photos-export...
Processed: 2268 photos, exported: 0, updated: 0, skipped: 0, updated EXIF data: 0, missing: 0, error: 2268
Elapsed time: 42.097 seconds
Lør 6 Feb 2021 07:00:53 UTC
Exporting 2268 photos to /Users/dsh/Pictures/Photos-export...
Processed: 2268 photos, exported: 0, updated: 0, skipped: 0, updated EXIF data: 0, missing: 0, error: 2268
Elapsed time: 52.717 seconds
Lør 6 Feb 2021 08:00:00 UTC
Exporting 2275 photos to /Users/dsh/Pictures/Photos-export...
Processed: 2275 photos, exported: 7, updated: 0, skipped: 0, updated EXIF data: 7, missing: 0, error: 2268
Elapsed time: 63.106 seconds
Lør 6 Feb 2021 09:00:00 UTC
Exporting 2268 photos to /Users/dsh/Pictures/Photos-export...
Processed: 2268 photos, exported: 0, updated: 0, skipped: 0, updated EXIF data: 0, missing: 0, error: 2268
Elapsed time: 34.738 seconds
`

This error seems to apply to each of them:

`Error exporting photo (AE8DFD44-648F-4415-9724-F7B4D2E33631: AE8DFD44-648F-4415-9724-F7B4D2E33631.JPG) as /Users/dsh/Pictures/Photos-export/Melissas album/AE8DFD44-648F-4415-9724-F7B4D2E33631.JPG: Invalid photo id: AE8DFD44-648F-4415-9724-F7B4D2E33631 (_photoinfo_export.pyc: 859)`

**Reply by RhetTbull** [reply]
What's the full command line you're using to export?

Are you using more than one Photos library?  This error looks like it's happening in the part of the export code that uses AppleScript to fetch missing photos from iCloud.  If the library open in Photos is not the one you're trying to export from, this error would get generated. If that's not the case I'll have to do some more digging.

**Reply by DanielHenriksen** [reply]
No, only one library. I (thought I) had the entire library locally and then there were much more errors. I then told Photos to keep them locally and tried again, but there are still errors (>2100) (and around 10 GBs). This may be videos - are videos a problem?

This is the code. I don't know if --only-new overrides --update, but that would make sense. I can remove the --download-missing as it should be redundant now.

`osxphotos export /Users/dsh/Pictures/Photos-export --directory '{folder_album}' --person-keyword --album-keyword --keyword-template '{created.year}' --download-missing --exiftool --update --only-new >> /Users/dsh/Pictures/Photos-export.log`

**Reply by RhetTbull** [reply]
I tried that exact combination of options and export worked as expected.  --only-new must be used with --update (otherwise you'll get an error) as --only-new assumes you are also doing an update to not re-export everything. 

osxphotos should export all media types (except for audio which aren't handled) so videos should work fine.  

If you're still seeing these errors it's likely Photos hasn't downloaded all the files from iCloud or this part of the code would never be reached.

Try adding the --use-photokit option.  This will use a native PhotoKit system framework to download the missing files instead of using AppleScript.  This is still an experimental feature but it seems to be more reliable than AppleScript for downloading the missing photos.  You'll be prompted by your terminal the first time you run this to allow access to the Photos library.

You might also try adding the --verbose option which will log extra details for each file being exported and that might be helpful to debug.

**Reply by DanielHenriksen** [reply]
Thank you! I'll try:)

**Reply by RhetTbull** [reply]
@DanielHenriksen are you still seeing these errors with export?  Did —use-photokit solve the problem?

**Reply by DanielHenriksen** [reply]
Yes, apparently. No errors :)
